### DISCUSSION TITLE
各ディレクトリ配下の index.dart に import を集約する方法について（Barrel ファイル）

### DISCUSSION THREAD

**Discussion by blendthink** [yumemi-inc/flutter-mobile-project-template#232] [General]
## メリット

- 別のディレクトリから参照してもいいクラスや関数だということが分かりやすい
- import する側の import 文が短くできる

## デメリット

- ディレクトリが増えるごとに index.dart も増える
- index.dart を作成しなければならない
- ルールを守っていない場合のカスタムリントを作成する必要がある
  - 別のディレクトリから扱うときには基本的にバレルを import するようにする
  - 別のディレクトリから扱いたいものだけバレルで export する

## 所感

Dart で monorepo 構成にするとパッケージ間で利用できるものはどちらにしてもまとめているため、そこまで大きなメリットにはならないかもしれません。自動生成やカスタムリントをどうするかが議論のポイントになりそうです。

↓社内での議論スレッド
https://yumemi.slack.com/archives/C0585H2BG9L/p1706506355658119

**Comment by tatsutakein** [comment]
最近は積極的に作成しなくてよいかも…と思い始めています．

理由としては，パッケージ内で利用する場合は大体近くに定義することがほとんどだと思うのと，パッケージ外で利用する場合は実質バレルのように export しているので…！
また，1 パッケージが極端に肥大化することはあまりないかなと思っております！

少なくとも現状ではメリットよりデメリットの方が上回っていそう…！という意見です 🙋

**Comment by K9i-0** [comment]
1. 個人的に感じたデメリットとして、パッケージで公開されているライブラリを知りたい時に、lib直下のファイルを見るだけだとわからずindex.dartを（場合によっては複数）辿らないといけないのが気になりました。

>import する側の import 文が短くできる

2. @tatsutakein さんと同じだと思うのですが、パッケージ内でのimportはマルチパッケージゆえにそこまで複雑な階層にならないのと、外部からはlib直下のファイルが実質バレルのようになるのであまりメリットになってないように感じます。

**Comment by blendthink** [comment]
お二人ともコメントありがとうございます🙏
では、Barrel ファイル形式をやめる方針で決定でお願いします💡
